src/main/scala/shine/DPIA/Compilation/StreamTranslation.scala

```diff
@@ -93,25 +137,81 @@ object StreamTranslation {
             arithexpr.arithmetic.RangeAdd(0, n, 1)))))))
 
     // OpenCL
-    case ocl.CircularBuffer(a, n, alloc, sz, dt1, dt2, load, input) =>
+    case ocl.CircularBuffer(a, n, alloc, size, dt1, dt2, load, input) =>
       val i = NatIdentifier(freshName("i"))
       str(input)(fun((i: NatIdentifier) ->:
         (expT(dt1, read) ->: (comm: CommType)) ->: (comm: CommType)
-      )(nextIn =>
-        oclI.CircularBufferI(a, n, alloc, sz, dt1, dt2,
-          fun(expT(dt1, read))(x =>
-            fun(accT(dt2))(o => acc(load(x))(o))),
-          nextIn, C)))
+      )(nextInput => {
+        if (!arithexpr.arithmetic.ArithExpr.isSmaller(size, alloc + 1).contains(true)) {
+          println(s"WARNING: circular buffer of size $alloc used for"
+            + s" a sliding window of size $size")
+        }
+
+        // TODO: generalize and make explicit? might not always be wanted
+        def allocGen(dt: DataType,
+                     C: (Phrase[AccType], Phrase[ExpType]) => Phrase[CommType]
+                    ): Phrase[CommType] = dt match {
+          case PairType(a, b) =>
+            allocGen(a, (wa, ra) => allocGen(b, (wb, rb) =>
+              C(UnzipAcc(alloc, a, b, PairAcc(alloc`.`a, alloc`.`b, wa, wb)),
+                Zip(alloc, a, b, read, ra, rb))
+            ))
+          case _ =>
+            shine.OpenCL.DSL.`new`(a)(alloc`.`dt, buffer => C(buffer.wr, buffer.rd))
+        }
+
+        allocGen(dt2, (bufWr: Phrase[AccType], bufRd: Phrase[ExpType]) => {
+          // TODO: unroll flags?
+          // prologue initialisation
+          forNat(n = size - 1, f = i => streamNext(nextInput, i, fun(expT(dt1, read))(x =>
+            acc(load(x))(bufWr `@` i)))) `;`
+            C(nFun(i =>
+              fun(expT(size`.`dt2, read) ->: (comm: CommType))(k =>
+                // load next value
+                streamNext(nextInput, i + size - 1, fun(expT(dt1, read))(x =>
+                  acc(load(x))(DropAcc(size - 1, n, dt2,
+                    CycleAcc(size - 1 + n, alloc, dt2, bufWr)) `@` i)
+                )) `;`
+                  // use neighborhood
+                  k(Drop(i, size, dt2,
+                    Cycle(i + size, alloc, dt2, bufRd)))
+              ),
+              arithexpr.arithmetic.RangeAdd(0, n, 1)
+            ))
+        })
+      }
+//        oclI.CircularBufferI(a, n, alloc, sz, dt1, dt2,
+//          fun(expT(dt1, read))(x =>
+//            fun(accT(dt2))(o => acc(load(x))(o))),
+//          nextIn, C)
```

Comment:
can be deleted?

Reply:
> Looks good to me, I added small comments. My only concern is that I believe the way you inlined some functions changes the translation evaluation order: before some variables would be translated as black-boxes before knowing their value, now their value is directly given to the translation. Do the changes have any impact on compilation time or generated code? It seems like this change already required adding a translation for `IdxDistribute` for example (or does this come from some other change?).

That is right. This changes the evaluation order. But this could easily be restored without the intermediate patterns, just be re-adding the DPIA functions that added the indirection.

I think that these functions were unnecessary, and we should always have supported `IdxDistribute` and  `Fst` + `Snd`. In fact, we should try to get to a design where we have exhaustive checks over the primitives as well ...

I'll fix the many small issues that you found. I should have made a cleanup path before opening the PR. Thanks for spotting them.